Refactor/repositories

FabianDobber opened the pull request:
### PR Description

This pull request introduces a comprehensive refactoring of the data persistence layer to improve architectural quality by adhering to the Single Responsibility Principle (SRP) and Separation of Concerns (SoC).

The core responsibility of mapping `IDataReader` objects to domain models has been extracted from the repositories and moved into dedicated `Mapper` classes.

#### Key Changes:

*   **Introduced Mappers**:
    *   A generic `IMapper<T>` interface has been created to define a standard contract for object mapping.
    *   Concrete mapper classes (e.g., `BoatMapper`, `ClientMapper`, `ReservationMapper`) have been implemented for each domain model, encapsulating all mapping logic.

*   **Simplified Repositories**:
    *   All repository classes (e.g., `BoatRepository`, `ClientRepository`) have been refactored. They are no longer responsible for mapping data records to models.
    *   Repositories now focus exclusively on their core responsibility: executing database queries (CRUD operations). They delegate the mapping of results to the appropriate `IMapper<T>` implementation, which is injected via dependency injection.

*   **Centralized Database Setup**:
    *   `IDbConnectionFactory` and `SqliteConnectionFactory` were introduced to centralize and manage the creation of database connections.
    *   `IDatabaseBootstrap`, `ISchemaInitializer`, `IDatabaseSeeder`, and `IDatabaseFixture` interfaces and their implementations were created to orchestrate the entire database setup (schema creation, seeding, and test data fixtures). This cleans up the application's entry point.

*   **Consolidated Helper Logic**:
    *   The duplicated `IsTableEmpty` method found in each seeder and fixture has been removed and replaced with a single, reusable extension method in the new `DbConnectionExtensions` class. This enforces the DRY principle and improves SRP for the seeder classes.

FabianDobber committed:
refactor(database): restructure repositories and centralize database setup

- Refactor repositories to follow SRP and SoC by introducing dedicated mappers
  and a generic mapping interface <T> for consistent data mapping.
- Add DbCommandExtensions.AddParameter method to simplify parameter addition
  to IDbCommand: creates a parameter, sets its name and value (uses DBNull.Value
  for nulls), and adds it to the command.
- Introduce SqliteDatabaseBootstrap to orchestrate database initialization:
  initializes schema, applies production seeders (Roles, BoatTypes, WindConstraints),
  and loads fixtures (dummy test data such as accounts).
- Add IDbConnectionFactory with SqliteConnectionFactory implementation to centralize
  connection creation.
- Introduce SqliteSchemaInitializer for single-step table creation.

FabianDobber committed:
fix(DI):
organize service registrations and remove duplicate entry
Removed a duplicate service registration for IDatabaseBootstrap.

FabianDobber committed:
refactor(data): centralize IsTableEmpty logic into an extension method
Removed the duplicated private `IsTableEmpty` method from all database seeder and fixture classes. This logic is now consolidated into a single `IsTableEmpty` extension method in `DbConnectionExtensions`.
This change eliminates duplication (DRY), and ensures the seeder/fixture classes better adhere to the Single Responsibility Principle.

FabianDobber committed:
refactor(data): make mappers robust by using named columns

Updated all data mapper classes to use name-based column lookups (`GetOrdinal`) instead of hardcoded numeric indexes.
Additionally, the database filename has been reverted to the original `Roeivereniging.db`.

Joppe20003 commented on ProjectBotenReservering.App/MauiProgram.cs:
```diff
@@ -34,11 +42,49 @@ public static MauiApp CreateMauiApp()
                 .Build();
 
+            string writableDirectory;
+            const string databaseName = "Roeivereniging.db";
```
why is this a const

Joppe20003 reviewed (approved):
seems looking good

FabianDobber commented on ProjectBotenReservering.App/MauiProgram.cs:
```diff
@@ -34,11 +42,49 @@ public static MauiApp CreateMauiApp()
                 .Build();
 
+            string writableDirectory;
+            const string databaseName = "Roeivereniging.db";
```
.db3 was a testing database which accidentally got pushed onto the development branch.

FabianDobber committed:
fix: fixed constructor ordering, and replaced all random names in BoatSeeder to "Remus Invictus"

Tein-dev reviewed (changes requested):
Good refactor. High quality code. Apply copilot's + my comments and I'll approve!!

FabianDobber committed:
fix:
Added SQL injection protection by making a whitelist hashset of possible schema names in DbConnectionExtension.cs
Fixed the column name in ClientManagementTaskMapper.cs ManagementTask_Id -> Management_Task_Id
Fixed the IsTableEmpty extension method to verify that the connection is open before executing the command, now opens as needed.
Fixed seeder order
Fixed an invalid time range in ReservationFixture.cs
Removed async keyword and changed the return type to Task.

FabianDobber committed:
fix: ClientFixture having random clubs, instead of Remus Invictus

FabianDobber committed:
fix: SqliteSchemaInitializer has been splitted into individual statements,
IDbConnectionFactory is now registered properly. Instead of using the new operator directly, it is registered with AddSingleton using a factory function, ensuring that if SqliteConnectionFactory implements IDisposable, it will be disposed correctly when the application shuts down.
IsTableEmpty now uses square brackets instead of double quotes

Tein-dev reviewed (approved):
I see that you've implemented the list with a foreach loop in the SqliteSchemaInitializer that we talked about for better error handling. LGTM🚀🚀🚀
